dist/datalayer.js

```diff
@@ -207,95 +208,161 @@ governing permissions and limitations under the License.
         if (!item.isValid()) {
             var message = 'The following item cannot be handled by the data layer ' +
                 'because it does not have a valid format: ' +
-                JSON.stringify(item);
+                JSON.stringify(item.getConfig());
             console.error(message);
             return;
         }
 
         var typeProcessors = {
             data: function(item) {
                 that._updateState(item);
-                that._triggerListeners(item.getConfig(), events.CHANGE);
+                that._triggerListeners(item);
             },
             event: function(item) {
-                that._triggerListeners(item);
                 if (item.getConfig().data) {
                     that._updateState(item);
-                    that._triggerListeners(item.getConfig(), events.CHANGE);
                 }
+                that._triggerListeners(item);
             },
             listenerOn: function(item) {
-                that._registerListener(item.getConfig());
+                that._processListenerOn(item);
             },
             listenerOff: function(item) {
-                that._unregisterListener(item.getConfig());
+                that._unregisterListener(item);
             }
         };
 
         typeProcessors[item.getType()](item);
     };
 
     /**
-     * Updates the state with the passed data configuration.
+     * Processes the item of type: listener on.
      *
-     * @param {DataConfig} item The data configuration.
+     * @param {DataLayer.Item} listener The listener.
      * @private
      */
-    DataLayer.Manager.prototype._updateState = function(item) {
-        DataLayer.utils.deepMerge(this._state, item.data);
+    DataLayer.Manager.prototype._processListenerOn = function(listener) {
+        var scope = listener.getConfig().scope;
```

Comment:
okay, rather than introducing `getIndex()` can't we simply read the index here before removing the listener from the stack?